src/lib/group.ts

```diff
@@ -0,0 +1,364 @@
+  }
+
+  #toAffine() {
+    return Pallas.fromAffine({
+      x: this.x.toBigInt(),
+      y: this.y.toBigInt(),
+      infinity: false,
+    });
+  }
+
+  /**
+   * Adds this {@link Group} element to another {@link Group} element.
+   *
+   * ```ts
+   * let g1 = Group({ x: 1, y: 1})
+   * let g2 = g1.add(g1)
+   * ```
+   */
+  add(g: Group) {
+    if (this.#isConstant() && g.#isConstant()) {
+      if (this.x.toBigInt() === 0n) {
+        return g;
+      } else if (g.x.toBigInt() === 0n) {
+        return this;
+      } else {
+        let g_proj = Pallas.add(this.#toAffine(), g.#toAffine());
+        return Group.#fromProjective(g_proj);
+      }
+    } else {
+      let [, x, y] = Snarky.group.add(this.#toTuple(), g.#toTuple());
+      return new Group({ x, y });
+    }
+  }
+
+  /**
+   * Subtracts another {@link Group} element from this one.
+   */
+  sub(y: Group) {
+    return this.add(y.neg());
+  }
+
+  /**
+   * Negates this {@link Group}. Under the hood, it simply negates the `y` coordinate and leaves the `x` coordinate as is.
+   */
+  neg() {
+    let { x, y } = this;
+    return new Group({ x, y: y.neg() });
+  }
+
+  /**
+   * Elliptic curve scalar multiplication. Scales the {@link Group} element `n`-times by itself, where `n` is the {@link Scalar}.
+   *
+   * ```typescript
+   * let s = Scalar(5);
+   * let 5g = g.scale(s);
+   * ```
+   */
+  scale(s: Scalar | number | bigint) {
+    let scalar =
+      typeof s === 'bigint' || typeof s === 'number'
+        ? Scalar.fromBigInt(BigInt(s))
+        : s;
+    let fields = scalar.toFields();
+
+    if (this.#isConstant() && fields.every((f) => f.isConstant())) {
+      let g_proj = Pallas.scale(this.#toAffine(), BigInt(scalar.toJSON()));
+      return Group.#fromProjective(g_proj);
+    } else {
+      let [, x, y] = Snarky.group.scale(this.#toTuple(), [
+        0,
+        ...fields.map((f) => f.value).reverse(),
+      ]);
+      return new Group({ x, y });
+    }
+  }
+
+  /**
+   * Assert that this {@link Group} element equals another {@link Group} element.
+   * Throws an error if the assertion fails.
+   *
+   * ```ts
+   * g1.assertEquals(g2);
+   * ```
+   */
+  assertEquals(g: Group, message?: string) {
+    let { x: x1, y: y1 } = this;
+    let { x: x2, y: y2 } = g;
+
+    x1.assertEquals(x2, message);
+    // y1.assertEquals(y2, message); need to enable this later on, but it breaks constraint backwards compatibility
+  }
+
+  /**
+   * Check if this {@link Group} element equals another {@link Group} element.
+   * Returns a {@link Bool}.
+   *
+   * ```ts
+   * g1.equals(g1); // Bool(true)
+   * ```
+   */
+  equals(g: Group) {
+    if (this.#isConstant() && g.#isConstant()) {
+      let { x: x1, y: y1 } = this;
+      let { x: x2, y: y2 } = g;
+
+      return Bool(x1.equals(x2).and(y1.equals(y2)));
+    } else {
+      /*
+    let equal_x = x1.equals(x2);
+    let equal_y = y1.equals(y2);
+    return equal_x.and(equal_y);*/
```

Comment:
this lgtm! if it results in a constraint change, I'd also try to switch the two equals constraints (it's OCaml so stuff is sometimes happening in reverse order :D)

Reply:
Ohh that was actually it, it was checking it in reverse

Reply:
Oh nope, this actually still behaves strangely :thinking: